packages/cli/src/cmds/beacon/initBeaconState.ts

```diff
@@ -13,10 +16,59 @@ import {
   initStateFromEth1,
 } from "@chainsafe/lodestar";
 import {getStateTypeFromBytes} from "@chainsafe/lodestar/lib/util/multifork";
+import {Checkpoint} from "@chainsafe/lodestar-types/phase0";
 import {downloadOrLoadFile} from "../../util";
 import {IBeaconArgs} from "./options";
 import {defaultNetwork, IGlobalArgs} from "../../options/globalOptions";
 import {getGenesisFileUrl} from "../../networks";
+import {WeakSubjectivityServers} from "../weakSubjectivityState";
+
+type WSResponse = {
+  current_epoch: number;
+  ws_checkpoint: string;
+  ws_period: number;
+  is_safe: boolean;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  ws_state: any;
+};
+
+type WeakSubjectivityData = {
+  state: TreeBacked<allForks.BeaconState>;
+  checkpoint: Checkpoint;
+};
+
+async function getWeakSubjectivityData(
+  config: IBeaconConfig,
+  args: IBeaconArgs & IGlobalArgs,
+  server: string,
+  logger: ILogger
+): Promise<WeakSubjectivityData> {
+  logger.info("Fetching weak subjectivity state from ChainSafe at " + server);
+  const response = await got(server, {searchParams: {checkpoint: args.weakSubjectivityCheckpoint}});
+  const responseBody = JSON.parse(response.body) as WSResponse;
+  const data = responseBody.ws_state.data;
+  const state = config.getForkTypes(data.slot).BeaconState.createTreeBackedFromJson(data, {case: "snake"});
+  if (!state) {
+    throw new Error("Weak subjectivity state not found for network " + args.network);
+  }
+  const checkpointData = (args.weakSubjectivityCheckpoint || responseBody.ws_checkpoint).split(":");
+  const checkpointRoot = checkpointData[0];
+  const checkpointEpoch = checkpointData[1];
+  const checkpoint = {root: fromHex(checkpointRoot), epoch: parseInt(checkpointEpoch)};
+  return {state, checkpoint};
+}
+
+async function initAndVerifyWeakSujectivityState(
```

Comment:
This is very weak verification, you can do much more

Reply:
What else would you like to see here?  The `isWithinWeakSubjectivityPeriod` function also does some root and epoch verification (i.e. it does not just check that the state is within the ws period).  That's just what they call the function in the spec.

Reply:
I think we need things from point 2 here: https://github.com/ethereum/eth2.0-specs/blob/dev/specs/phase0/weak-subjectivity.md#weak-subjectivity-sync-procedure

In the spec where it says `store`, it means the latest state stored in our state archive (which may or may not exist).

For the first bullet point:
> IF epoch_number > store.finalized_checkpoint.epoch, then ASSERT during block sync that block with root block_root is in the sync path at epoch epoch_number. Emit descriptive critical error if this assert fails, then exit client process.

We will probably need a new option to pass to the beacon node on init so that during backfill, this check can happen.

The second bullet point:
> IF epoch_number <= store.finalized_checkpoint.epoch, then ASSERT that the block in the canonical chain at epoch epoch_number has root block_root. Emit descriptive critical error if this assert fails, then exit client process.

That can be checked here.

Reply:
Again, `isWithinWeakSubjectivityPeriod` does those checks here:

https://github.com/ChainSafe/lodestar/pull/2571/files/3e3a81a2818e81929d9cbf1a471ef32713ccdce4#diff-51a07be7e62b602a2219ab9614f4e88b9fb63ab0d590886478c6504120349edeR73

However, as discussed in your other comment, it does not use the `store` because, as it currently stands, we implicitly don't have an existing store when fetching from wsState because if we have something stored in the db, then we init from that and not from a wsState.

Reply:
@wemeetagain  Maybe we can just fetch from the db and then if we also have the `fetchWS` flag enabled, then just use the db from state as the `store` for `isWithinWeakSubjectivityPeriod`?

Reply:
Oh, I see.
I think we'll want to allow ws sync even when a state exists in the db. The thing is, the long-range attack can happen whether or not the user already has a stored state or not. So if the user explicitly enables ws sync, it should be able to fetch/use the latest ws state.

As far as the cli goes, we can have:
- `wsStateFile`: string, default to undefined - if set, use this state for ws sync
- `wsCheckpoint`: string, default to undefined - if set, use this checkpoint for safety checks and (possibly) for fetching the ws state
- `wsServerUrl`: string, default to undefined - if undefined and ws sync is enabled and a state file is not provided (whether the user provides a ws checkpoint or not), use our provided infura url
- `wsSyncLatest`: boolean, default to false - if true, and `wsStateFile` and `wsCheckpoint` are not defined, fetch the latest ws checkpoint from infura, use that for safety check and fetching the ws state

so if `wsStateFile`, `wsCheckpoint`, or `wsSyncLatest` are set, the chain will be initialized with a ws state as the anchor state. Depending on how the db is populated, additional safety checks would be performed.

if `wsCheckpoint` or `wsSyncLatest` are set, `wsServerUrl` is used to fetch a ws state.

if `wsStateFile` and `wsCheckpoint` are set, `wsCheckpoint` is used for safety checks even tho no ws state is fetched via url.

Reply:
@wemeetagain pushed up this ^